feat(patient group) bulk assignment from the patient registry

jeremielodi opened the pull request:
closes #3519

jeremielodi committed:
feat(patient group) bulk assingenment from the registry

jniles commented on client/src/modules/patients/patients.service.js:
```diff
@@ -128,6 +129,15 @@ function PatientService(
       .then(service.util.unwrapHttpResponse);
   }
 
+  function bulkUpdateGroups(patientUuids, subscribedGroups) {
```
Nice naming scheme :+1:

jniles commented on client/src/modules/patients/registry/editPatientGroup.js:
```diff
@@ -0,0 +1,50 @@
+    } else {
+      const subscribedGroups = getGroups();
+      if (subscribedGroups.length > 0) {
+        Patient.bulkUpdateGroups(data, subscribedGroups)
```
If you do a `return Patient.bulkUpdateGroups()`, you could benefit from the loading indicator on the `bh-loading-button`.

jniles commented on client/src/modules/patients/registry/editPatientGroup.js:
```diff
@@ -0,0 +1,50 @@
+  function getGroups() {
+    const subscribedGroups = [];
+    const keys = Object.keys(vm.subscribedGroups);
+    keys.forEach(groupUuid => {
```
Just FYI - angular has a shortcut for this.  Instead of doing:
```js
const keys = Object.keys(someObject);
keys.forEach(key => { 
  console.log(someObject[key]);
});
```
you can do the same thing in a single line:
```js
angular.forEach(someObject, (value, key) => {
  console.log(key, value);
});
```

I personally don't really care which is used, but figured I would mention it in case you didn't know!

jniles commented on client/src/modules/patients/registry/registry.js:
```diff
@@ -268,6 +268,25 @@ function PatientRegistryController(
+      return;
+    }
+    const data = [];
+    patients.forEach(p => {
```
This could be a map!
```js
const data = patients.map(p => p.uuid);
```

jniles commented on server/controllers/medical/patients/groups.js:
```diff
@@ -114,3 +114,39 @@ function update(req, res, next) {
+  const groups = [].concat(subscribedGroups);
+
+  // Clear assigned groups
+  const removeAssignmentsQuery = 'DELETE FROM patient_assignment WHERE patient_uuid = ?';
```
I mentioned this to you in person, but it would be great to add an option to skip the "delete" request on the client.  This way would could support both:
 1. Replace all groups by deleting the old values OR
 2. Appending new values to the patients group list

jniles commented on server/controllers/medical/patients/groups.js:
```diff
@@ -114,3 +114,39 @@ function update(req, res, next) {
+      const assignment = {
+        uuid : db.uuid(),
+        patient_uuid : db.bid(patientUuid),
+        patient_group_uuid : db.bid(groupUuid),
```
Great variable naming!  This code is super clear.

jniles reviewed (commented):
@jeremielodi This is a super good pull request.  The code is clear, the variable names are well formed.  It doesn't contain any other changes besides this feature, so it is super focused.  Excellent work!

I've made a few comments below and in person.

One thing we probably should have is an integration and end to end test to make sure that these don't break in the future. 👍

jeremielodi committed:
bulk patient group assignment e2e test

mbayopanda commented on client/src/modules/patients/registry/editPatientGroup.html:
```diff
@@ -0,0 +1,51 @@
+  data-modal="patient-search"
+  novalidate>
+
+  <div class="modal-header search-modal-header">
```
Is it possible to remove the class `search-modal-header` here

mbayopanda commented on client/src/modules/patients/registry/editPatientGroup.html:
```diff
@@ -0,0 +1,51 @@
+    </ol>
+  </div>
+
+  <div class="modal-body search-modal">
```
Is it also possible to remove the class `search-modal-header` here

mbayopanda commented on client/src/modules/patients/registry/editPatientGroup.html:
```diff
@@ -0,0 +1,51 @@
+  </div>
+
+  <div class="modal-body search-modal">
+    <div class="col-sm-12">
```
This `<div class="col-sm-12">` is not necessary, could you remove it

mbayopanda commented on client/src/i18n/fr/form.json:
```diff
@@ -302,6 +302,7 @@
             "CASHBOX": "Caisse",
             "CASHIER": "Caissier",
             "CHECK_ALL":"Tout cocher",
+            "CHANGE_PATIENT_GROUP":"Changer group de patient",
```
`Changer groupe de patient`

mbayopanda commented on client/src/i18n/fr/patient_reg.json:
```diff
@@ -6,6 +6,7 @@
     "PATIENT_DETAILS": "Détails Patient",
     "SUCCESS": "Patient enregistré avec succès",
     "IDENTITY": "Identité du patient",
+    "REMOVE_PREVIOUS_ASSIGNMENT" : "Supprimer tous les groups déjà assignés à ces patients",
```
Supprimer tous les groupes déjà assignés à ces patients

mbayopanda commented on client/src/i18n/en/patient_reg.json:
```diff
@@ -6,6 +6,7 @@
     "SUCCESS": "Patient Registered Successfully",
     "CARD": "Patient Card",
     "IDENTITY": "Patient Identity",
+  "REMOVE_PREVIOUS_ASSIGNMENT" : "Remove all aleady assigned groups to this patients",
```
`Delete all groups already assigned to these patients`

mbayopanda reviewed (commented):
I made some comments

mbayopanda commented on client/src/modules/patients/registry/editPatientGroup.html:
```diff
@@ -0,0 +1,51 @@
+        <div class="checkbox">
+          <label>
+            <input type="checkbox" ng-true-value="1" ng-false-value="0" ng-model="$ctrl.removeAssignedGroups">
+            <span translate>PATIENT_REG.REMOVE_PREVIOUS_ASSIGNMENT</span>
```
🐛 the option of removing all previous group assignment doesn't work 🐛

mbayopanda reviewed (commented):
The option of removing all previous group assignments doesn't work

jniles committed:
Merge remote-tracking branch 'upstream/master' into patient_group_assign_bulk

jniles committed:
chore(patients): fix-up bulk group assignment

Fixes review comments and other improvements to bring bulk assignment up
to date.

jniles committed:
chore: fix up tests for newer protractor

jniles committed:
fix(patients): make merge functionality work

jniles committed:
fix(patients): ensure patients are deselected

Fixes a broken test where patients were not deselected after their
group changes.

jniles committed:
fix(patients): remove all patient groups submits

Removing all patient groups from a patient didn't work previously.  This
was because we had an invalid check in the submit method that would
never send the request to the server.  Now, we check that either there
are groups to submit or the user is removing all groups.

We also ensure that if the user is removing all groups, the checkboxes
are cleared and vice versa.
